utils/service_discovery/abstract_config_store.py

```diff
@@ -34,6 +34,125 @@ class KeyNotFound(Exception):
     pass
 
 
+class _TemplateCache(object):
+    """
+    Store templates coming from the configuration store and files from auto_conf.
+
+    Templates from different sources are stored in separate attributes, and
+    reads will look up identifiers in both of them in the right order.
+
+    read_func is expected to return raw templates coming from the config store.
+
+    The cache must be invalidated when an update is made to templates.
+    """
+
+    def __init__(self, read_func, root_template_path):
+        self.read_func = read_func
+        self.root_path = root_template_path
+        self.kv_templates = defaultdict(lambda: [[]] * 3)
+        self.auto_conf_templates = defaultdict(lambda: [[]] * 3)
+        self._populate_auto_conf()
+
+    def invalidate(self):
+        """Clear out the KV cache"""
+        log.debug("Clearing the cache for configuration templates.")
+        self.kv_templates = defaultdict(lambda: [[]] * 3)
+
+    def _populate_auto_conf(self):
+        """Retrieve auto_conf templates"""
+        raw_templates = get_auto_conf_images(full_tpl=True)
+        for image, tpls in raw_templates.iteritems():
+            for check_name, init_tpl, instance_tpl in zip(*tpls):
+                if image in self.auto_conf_templates:
+                    if check_name in self.auto_conf_templates[image][0]:
+                        log.warning("Conflicting templates in auto_conf for image %s and check %s. "
+                                "Please check your template files." % (image, check_name))
+                        continue
+                    self.auto_conf_templates[image][0].append(check_name)
```

Comment:
Oh I get it, we stick to lists for "compatibility" across the board when dealing with the configs. I think we could maybe refactor in the future, but it's a legit choice.